### ISSUE TITLE
ORM - Save NULL foreign keys

### ISSUE THREAD

**Issue by calinrada** [phalcon/cphalcon#2039]
I have a table with a froiegn key (scan_id) that can be null. In the model i have: 

```php
    public function beforeValidation()
    {
        if ($this->scan_id == '') {
            $this->scan_id = new \Phalcon\Db\RawValue('NULL');
        }
    }
```

But when i try to save the object i get "Value of field "scan_id" does not exist on referenced table". 

Any suggestions ? Thanks.

Phalcon 1.2.6
PHP 5.5.8

**Comment by DusanBrejka** [comment]
Any updates on this issue, guys? 

I really wanted to keep all the real valid DB relationships inside my models, so just turning off foreignKey check and handling it myself for every models was not an option. I came up with this temporary shitty solution that overrides \Phalcon\Mvc\Model\Manager::getBelongsTo and return foreignKey = true only if there is any non-null value set for selected attribute. 

getBelongsTo method is called in _checkForeignKeysRestrict method in \Phalcon\Mvc\Model right before it tries to create/update corresponding rows.

```php
$di->set('modelsManager', function() {
    return new \ModelManager();
}, true);
```

```php
class ModelManager extends Phalcon\Mvc\Model\Manager {
    public function getBelongsTo($model) {
        /** @var \Phalcon\Mvc\Model $model */
        if(!empty( $originalBelongsTo = parent::getBelongsTo($model) )) {
            $nullFields = array_diff(
                $model->getModelsMetaData()->getAttributes($model),
                $model->getModelsMetaData()->getNotNullAttributes($model)
            );
            if(!empty($nullFields)) {
                /** @var \Phalcon\Mvc\Model\Relation $oBT */
                foreach($originalBelongsTo as &$oBT) {
                    /** @var \Phalcon\Db\RawValue $rawValue */
                    if( is_string($field = $oBT->getFields()) &&
                        in_array($field, $nullFields) &&
                        ( $model->$field === null || (( $rawValue = $model->$field) instanceof \Phalcon\Db\RawValue && trim(strtolower($rawValue->getValue())) == 'null' ) ) ) {
                        $oBT = new Phalcon\Mvc\Model\Relation(
                            $oBT->getType(),
                            $oBT->getReferencedModel(),
                            $oBT->getFields(),
                            $oBT->getReferencedFields(),
                            [ 'foreignKey' => false ] + $oBT->getOptions()
                        );
                    }
                }
            }
        }
        return $originalBelongsTo;
    }
}
```

**Comment by GBraL** [comment]
Hi friends,

When i try use this solution i receive a message saying not access the private attribute.

I use private attributes and create gets and setters to all attributes.

The error ocured on this line: $model->$field === null

How i resolve this?

**Comment by dugwood** [comment]
You should try to use the getter, such as `$model->{'get'.$field}() === null` or if it doesn't work (I didn't test it) try setting `$getter = 'get'.$field';` then use it with `$model->$getter() === null`

**Comment by GBraL** [comment]
I thought about it but the problem with this solution is that some use gets setters and with different attribute name.

By this I would have to treat each case individually or standardize all my models.
It is striking that the Phalcon can not handle such a situation.

I can not insert NULL value in a FK because there is a relationship with the parent table and I can not remove the relationship is not adding any value that does not exist in the child table will not be treated by the application, launching a database exception.

**Comment by stibiumz** [comment]
My solution https://gist.github.com/stibiumz/7d92ed9ebb33ff67d656, I've not tried with multiple referenced fields nor conditions, but should work

**Comment by GBraL** [comment]
Hi my dear.

I have 2 FKs. One or other is not null.

The sollution i have found to this is disable the foreign key check, removing "'foreignKey' => true" and making validation in beforeValidation function.

I try use your sollution has work, but if i try insert a value not exists on reference table your solution not work fine. The next error is show:

"Fatal error: Uncaught exception 'PDOException' with message 'SQLSTATE[23503]: Foreign key violation: 7 ERROR: insert or update on table..."

**Comment by stibiumz** [comment]
@GBraL sorry, a typo modifying a part when I was pasting the gist https://gist.github.com/stibiumz/7d92ed9ebb33ff67d656/revisions

**Comment by GBraL** [comment]
Thank you my dear.

Works fine.

I only need add the validation when either FK is mandatory.

Regards.

**Comment by andresgutierrez** [comment]
This is implemented in 2.0.3:

```php
class RobotsParts extends Model
{
    public function initialize()
    {
        $this->belongsTo("robots_id", "Some\Robots", "id", array(
            "foreignKey" => [
                "message" => "Robots is required",
                "allowNulls" => true
            ]
        ));
    }
}
```

**Comment by ianhaycox** [comment]
Note: The documentation at https://docs.phalconphp.com/en/latest/reference/models.html#virtual-foreign-keys is incorrect. allowNull does not work, but allowNulls (@andresgutierrez) does work.
